### PR TITLE
honor requests_max based on cgroup_limits if configured

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
honor requests_max based on cgroup_limits if configured

## Motivation and Context
container limits would not be properly honored in
our current implementation, mem.VirtualMemory()
function only reads /proc/meminfo which points to
the host system information inside the container.

## How to test this PR?
Run MinIO with proper memory limits such as 

```
~ docker run --rm --memory 2gb minio/minio server /tmp/xl{1...4}
```

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation updated
- [ ] Unit tests added/updated

### CODE REVIEW

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

File: cmd/handler-api.go [hunk 1]
```diff
@@ -48,6 +51,40 @@ type apiConfig struct {
...
+		}
+		available = cgroupLimit
+	}
+
+	// No limit set, It's the highest positive signed 64-bit
+	// integer (2^63-1), rounded down to multiples of 4096 (2^12),
+	// the most common page size on x86 systems - for cgroup_limits.
+	if available != 9223372036854771712 {
+		// This means cgroup memory limit is configured.
+		return
```
Comment (vadmeste) [inline cmd/handler-api.go]: We are returning early here with available = 8 Gb when the OS is not 'linux'

File: cmd/handler-api.go [hunk 2]
```diff
@@ -48,6 +51,40 @@ type apiConfig struct {
 	deleteCleanupInterval       time.Duration
 }
 
+const cgroupLimitFile = "/sys/fs/cgroup/memory/memory.limit_in_bytes"
```
Comment (vadmeste) [inline cmd/handler-api.go:54]: Is this file old enough ? Not able to find when it is introduced

**Review by vadmeste — COMMENTED** [review]
Some comments

File: cmd/handler-api.go (hunk 1 above)
Comment (klauspost) [reply cmd/handler-api.go]: Yeah. move this into the linux section.

File: cmd/handler-api.go [hunk 3]
```diff
@@ -48,6 +51,40 @@ type apiConfig struct {
...
+
+func availableMemory() (available uint64) {
+	available = 8 << 30 // Default to 8 GiB when we can't find the limits.
+
+	if runtime.GOOS == "linux" {
```
Comment (klauspost) [inline cmd/handler-api.go:71]: (github UI bugging out)
```suggestion
	for runtime.GOOS == "linux" {
		buf, err := ioutil.ReadFile(cgroupLimitFile)
		if err != nil {
			break
		}
		cgroupLimit, err := strconv.ParseUint(string(buf), 10, 64)
		if err != nil {
			break
		}
		// No limit set, It's the highest positive signed 64-bit
		// integer (2^63-1), rounded down to multiples of 4096 (2^12),
		// the most common page size on x86 systems - for cgroup_limits.
		if cgroupLimit == 9223372036854771712 {
			// no-limit set proceed to set the limits based on virtual memory.
			break
		}
		return cgroupLimit
	}
```

**Review by klauspost — CHANGES_REQUESTED** [review]

File: cmd/handler-api.go (hunk 2 above)
Comment (harshavardhana) [reply cmd/handler-api.go:54]: This present since cgroups was implemented

**Review by klauspost — APPROVED** [review]

**Review by vadmeste — APPROVED** [review]
LGTM

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Pull request merged**